autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/impl/importing/gpm/webflow/ImportGooglePasswordsWebFlowViewModel.kt

```diff
@@ -0,0 +1,131 @@
+
+    fun onPageStarted(url: String?) {
+        Timber.i("onPageStarted: $url")
+    }
+
+    fun onPageFinished(url: String?) {
+        Timber.i("onPageFinished: $url")
+    }
+
+    suspend fun onCsvAvailable(csv: String) {
+        when (val parseResult = csvPasswordConverter.readCsv(csv)) {
+            is CsvPasswordImportResult.Success -> onCsvParsed(parseResult)
+            is CsvPasswordImportResult.Error -> onCsvError()
+        }
+    }
+
+    private suspend fun onCsvParsed(parseResult: Success) {
+        val jobId =
+            passwordImporter.importPasswords(parseResult.loginCredentialsToImport, parseResult.numberPasswordsInSource)
+        val resultBundle = Bundle().also {
+            it.putParcelable(
+                RESULT_KEY_DETAILS,
+                ImportGooglePasswordResult.Success(
+                    importedCount = parseResult.loginCredentialsToImport.size,
+                    foundInImport = parseResult.loginCredentialsToImport.size,
+                    importJobId = jobId,
+                ),
+            )
+        }
+        _viewState.value = ViewState.UserFinishedImportFlow(resultBundle)
+    }
+
+    fun onCsvError() {
+        Timber.e("cdr Error decoding CSV")
+        val resultBundle = Bundle().also {
+            it.putParcelable(RESULT_KEY_DETAILS, ImportGooglePasswordResult.Error)
+        }
+        _viewState.value = ViewState.UserFinishedImportFlow(resultBundle)
+    }
+
+    fun onCloseButtonPressed(url: String?) {
+        if (url?.startsWith(ENCRYPTED_PASSPHRASE_ERROR_URL) == true) {
+            val resultBundle = Bundle().also {
+                it.putParcelable(RESULT_KEY_DETAILS, ImportGooglePasswordResult.Error)
+            }
+            _viewState.value = ViewState.UserFinishedImportFlow(resultBundle)
+        } else {
+            terminateFlowAsCancellation(url ?: "unknown")
+        }
+    }
+
+    fun onBackButtonPressed(
+        url: String?,
+        canGoBack: Boolean,
+    ) {
+        // if WebView can't go back, then we're at the first stage or something's gone wrong. Either way, time to cancel out of the screen.
+        if (!canGoBack) {
+            terminateFlowAsCancellation(url ?: "unknown")
+            return
+        }
+
+        _viewState.value = ViewState.NavigatingBack
+    }
+
+    private fun terminateFlowAsCancellation(stage: String) {
+        _viewState.value = ViewState.UserCancelledImportFlow(stage)
+    }
+
+    sealed interface ViewState {
+        data object Initializing : ViewState
+        data class UserCancelledImportFlow(val stage: String) : ViewState
+        data class UserFinishedImportFlow(val bundle: Bundle) : ViewState
+        data object NavigatingBack : ViewState
+    }
+
+    sealed interface BackButtonAction {
+        data object NavigateBack : BackButtonAction
+    }
+
+    companion object {
+        private const val ENCRYPTED_PASSPHRASE_ERROR_URL = "https://passwords.google.com/error/sync-passphrase"
```

Comment:
what's the intention behind the stage and this url? are we trying to understand where the user leaves the flow and why?